Adding scenario for multinode job array with advance reservation

Mahalty opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Description

- Automate following scenario - "multinode job array with advance reservation"

#### Attach Test and Valgrind Logs/Output
[execution_logs.txt](https://github.com/PBSPro/pbspro/files/3202408/execution_logs.txt)

<!--- Pull Request Guidelines: [Pull Request Guidelines](https://pbspro.atlassian.net/wiki/spaces/DG/pages/1187348483/Pull+Request+Guidelines) -->

bhroam commented on test/tests/functional/pbs_reservations.py:
```diff
@@ -883,3 +883,80 @@ def test_fail_confirm_resv_message(self):
+        Test multinode job array with advance reservation
+        """
+        if (len(self.moms) < 2):
+            self.skip_test("Test requires 2 moms as input")
```
Say "Test requires 2 moms: use -p mom1:mom2:

bhroam commented on test/tests/functional/pbs_reservations.py:
```diff
@@ -883,3 +883,80 @@ def test_fail_confirm_resv_message(self):
+        self.server.expect(JOB, {'job_state=Q': 3}, count=True,
+                           extend='t', id=jid)
+        self.server.expect(JOB, {'job_state': 'Q'}, id=subjid[3])
+        self.server.deljob(subjid[3])
```
I think self.server.delete() is the right PTL interface to use. (and other places in the test plan)

bhroam commented on test/tests/functional/pbs_reservations.py:
```diff
@@ -883,3 +883,80 @@ def test_fail_confirm_resv_message(self):
+                 'Resource_List.select': '2:ncpus=1'}
+        j1 = Job(PBSROOT_USER)
+        j1.set_sleep_time(60)
+        j1.set_attributes(attrs)
```
Consider passing attrs to Job().  it'll save you the call to set_attributes().

Mahalty commented:
Latest execution-logs:
[Latest_execution_logs.txt](https://github.com/PBSPro/pbspro/files/3206049/Latest_execution_logs.txt)

Klovely commented on test/tests/functional/pbs_reservations.py:
```diff
@@ -883,3 +883,80 @@ def test_fail_confirm_resv_message(self):
+        self.server.expect(JOB, {'job_state': 'S'}, id=subjid2[0])
+        self.server.sigjob(jobid=subjid2[1], signal="suspend")
+        self.server.expect(JOB, {'job_state': 'S'}, id=subjid2[1])
+        self.server.deljob(subjid2[1])
```
Remove this line .
Instead of deleting  subjid2[1] , check subjid2[2] and subjid[3] started running.

Klovely commented on test/tests/functional/pbs_reservations.py:
```diff
@@ -883,3 +883,80 @@ def test_fail_confirm_resv_message(self):
+        self.server.expect(JOB, {'job_state=Q': 3}, count=True,
+                           extend='t', id=jid)
+        self.server.expect(JOB, {'job_state': 'Q'}, id=subjid[3])
+        self.server.deljob(subjid[3])
```
As we discussed, even if we delete this line, next job-array would be in Q and no effect in test case objective.
So I would suggest remove this line.

Klovely replied:
Generic comment  could you please add doc string to increase readability of test steps.

Mahalty commented:
Latest execution-logs:
[Latest_execution_logs_1.txt](https://github.com/PBSPro/pbspro/files/3207378/Latest_execution_logs_1.txt)

bhroam commented on test/tests/functional/pbs_reservations.py:
```diff
@@ -914,46 +916,60 @@ def test_advance_resv_with_multinode_job_array(self):
         for i in range(1, 6):
             subjid.append(j.create_subjob_id(jid, i))
-
+        # Wait for reservation to be in running state
```
Turn this comment into a self.logger.info().  It is nicer for the test runner to know why they are waiting 30s.  It also serves the same purpose as the comment.

bhroam commented on test/tests/functional/pbs_reservations.py:
```diff
@@ -914,46 +916,60 @@ def test_advance_resv_with_multinode_job_array(self):
         self.server.expect(JOB, {'job_state': 'R'}, id=subjid2[2])
+        self.server.expect(JOB, {'job_state': 'R'}, id=subjid2[3])
+        # Delete running jobs
+        self.server.delete(subjid2[2], subjid2[3])
```
You need to pass the jobs in as a list.  This passes them in as two different arguments.  The second argument to delete() is the extend field.  So make this self.server.delete([subjob[2], subjob[3]])

bhroam commented on test/tests/functional/pbs_reservations.py:
```diff
@@ -914,46 +916,60 @@ def test_advance_resv_with_multinode_job_array(self):
-                           extend='x', offset=120)
+        self.server.sigjob(subjid2[1], 'resume')
+        self.server.expect(JOB, {'job_state=R': 2}, count=True,
+                           extend='t', id=jid2)
```
Resuming subjobs 1 and 2 will not resume them immediately.  It will change their substate from 43 to 45.  A job in substate 45 is looked at by the scheduler to resume it when there is space to do so.  If I understand properly,  there should be 3 running jobs and 2 suspended jobs right now.  What this does is wait for the 3 running jobs to end, and then the 2 suspended jobs resume.  Is this what you want?

bhroam commented:
Some of your new comments are explain what you are doing and not why.  The code nicely explains what you are doing.  Saying "See job array in the B state" and then immediately following it by an expect to see it in the B state doesn't really help.  Add a comment if you need to explain why you are doing it.  If a code reader sees a few comments that explain what the code is doing, they'll stop reading comments.  They'll just look at the code.  Consider rephrasing some of your comments or removing some.

Mahalty commented:
Latest execution-logs:
[Latest_execution_logs_2.txt](https://github.com/PBSPro/pbspro/files/3210783/Latest_execution_logs_2.txt)

Klovely commented:
Latest changes looks good to me .

bhroam commented on test/tests/functional/pbs_reservations.py:
```diff
@@ -1074,22 +1070,18 @@ def test_advance_resv_with_multinode_job_array(self):
                            extend='t', id=jid2)
         self.server.expect(JOB, {'job_state': 'F'},
-                           id=jid, extend='x', offset=180)
+                           id=jid2, extend='x', offset=180)
```
Consider adding a self.logger.info() explaining what the 120s sleep is for.  It'd be nicer to the test runner to know why they are waiting.  It isn't necessary though.

Mahalty commented:
Thanks @bhroam  for the review. I have addressed your review comment

@PBSPro/pbspro-maintainers I have squashed commits and rebased my branch. Please take this PR forward
